designs/2021-stricter-rule-test-validation/README.md

```diff
@@ -0,0 +1,286 @@
+- The maintenance burden is higher as changes to the test case need to be replicated in both `code` and `output`
+- The test case is less concise, needlessly increasing the size/length of the test file
+
+So we will disallow repeating the test case `code` as `output` for non-fixable test cases using a new assertion:
+
+```pt
+AssertionError [ERR_ASSERTION]: Test error object 'output' matches 'code'. Omit 'output' or use `output: null` for non-fixable test cases.
+```
+
+And similarly, suggestions must actually produce a change:
+
+```pt
+AssertionError [ERR_ASSERTION]: Test suggestion object 'output' matches 'code' and is thus a no-op.
+```
+
+To clearly and concisely indicate that a test case produces no autofix, it is recommended to omit the `output` property entirely. Note that we will also allow `output: null` or `output: undefined` which can be useful when dynamically generating test cases (e.g. `output: hasAutofix ? autofixedCode : null`), and because many existing test cases are written with `output: null` as that used to be necessary to assert that the test case had no autofix (it isn't necessary anymore as of [ESLint v7](https://eslint.org/docs/user-guide/migrating-to-7.0.0#additional-validation-added-to-the-ruletester-class)).
+
+An existing lint rule [eslint-plugin/prefer-output-null](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/prefer-output-null.md) enforces the desired behavior and is autofixable which should help with migration.
+
+### Disallow identical test cases
+
+Duplicate test cases can cause confusion, can be hard to detect manually in a long file, and serve no purpose.
+
+We will add an assertion when a duplicate test case is detected:
+
+```pt
+AssertionError [ERR_ASSERTION]: Detected duplicate test case.
+```
+
+As test cases end up represented as objects, the problem here is essentially to detect duplicate objects in an array. To handle this, here's one efficient algorithm (also mentioned in this [article](https://medium.com/programming-essentials/how-to-know-if-an-array-has-duplicates-in-javascript-27d99ab9a0d2)):
+
+1. Maintain a set of the test cases we have seen so far.
+2. For each test case, use our existing dependency [json-stable-stringify-without-jsonify](https://www.npmjs.com/package/json-stable-stringify-without-jsonify) to get a string representation of the test case object.
+   - This library handles deep-sorting of object keys, unlike `JSON.stringify()`.
+   - We need to skip test cases that contain non-serializable properties like functions or RegExp objects. Since rule options can contain any number of non-serializable properties, we will just skip all test cases with rule options. Detecting duplicates isn't critical, so it's okay that we will skip some test cases, but we could potentially make this smarter later.
```

Comment:
There are other possibly non-serializable properties that can appear in test cases, as config properties.

Currently, I think it's `settings` and `parserOptions` (depending on the parser) as possibly unserializable. With the [flat config](https://github.com/eslint/eslint/blob/18f5e05bce10503186989d81ca484abb185a2c9d/lib/config/flat-config-schema.js#L437-L457), we'll have `settings`, `languageOptions` (`parser` can be a parser object with `parse` / `parseForESlint` function) and `plugins` as possibly unserializable properties.

Reply:
Good call out, I'm happy to limit this to only test cases which only contain known-serializable properties.